source/util.d

```diff
@@ -1,33 +1,131 @@
 module util;
 
+import core.attribute : mustuse;
+import std.algorithm : map;
+import std.array : join;
+import std.format : format;
+import std.meta : All = allSatisfy, IndexOf = staticIndexOf, Map = staticMap;
+import std.range : iota;
 import std.stdio : File, stderr, stdout;
+import std.sumtype;
 
-/**
-Creates an annonymous enum that extends from T and exposes aliases that bind
-to each enum member of T.
+@mustuse struct Result(Success, Error)
+{
+	@safe pure nothrow @nogc:
+
+	SumType!(Success, Error) st;
+	alias st this;
+
+	static Ok(Success value)
+	{
+		Result self = { st: value };
+		return self;
+	}
 
-Params:
-    T = enum to bind the exposed aliases.
+	static Err(Error value)
+	{
+		Result self = { st: value };
+		return self;
+	}
 
-Note: This only binds to T. If the underlying type of T is another enum, those
-members are not aliased.
+	bool opCast(T : bool)() const
+	{
+		// access private data of SumType to avoid 'match'
+		enum stringify(alias member) = member.stringof;
+		enum tid = IndexOf!(Success, st.Types);
+		const tag = st.tupleof[IndexOf!("tag", Map!(stringify, st.tupleof))];
+		return tag == tid;
+	}
+
+	Success success() @system const
```

Comment:
These should probably `inout`. There might be types with mutable members or non-const indirection that wouldn't work generically, because `const` is enforced.